### ISSUE TITLE
[spirv] 'memref.alloc' op unexpected allocation without upper bound shapes

### ISSUE THREAD

**Issue by vivekkhandelwal1** [iree-org/iree#14314]
### What happened?

Hi, I'm getting the following error during the compilation of an IR:
```
<stdin>:810:21: error: 'memref.alloc' op unexpected allocation without upper bound shapes
<stdin>:26:3: note: called from
<stdin>:820:10: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"vulkan", "vulkan-spirv-fb", {spirv.target_env = #spirv.target_env<#spirv.vce<v1.6, [Shader, Float64, Float16, Int64, Int16, Int8, StorageBuffer16BitAccess, StorageUniform16, StoragePushConstant16, StorageBuffer8BitAccess, UniformAndStorageBuffer8BitAccess, StoragePushConstant8, GroupNonUniform, GroupNonUniformVote, GroupNonUniformArithmetic, GroupNonUniformBallot, GroupNonUniformShuffle, GroupNonUniformShuffleRelative, GroupNonUniformClustered, GroupNonUniformQuad, VariablePointers, VariablePointersStorageBuffer, DotProduct, DotProductInputAll, DotProductInput4x8BitPacked, DotProductInput4x8Bit, CooperativeMatrixNV], [SPV_KHR_16bit_storage, SPV_KHR_8bit_storage, SPV_KHR_storage_buffer_storage_class, SPV_KHR_variable_pointers, SPV_NV_cooperative_matrix]>, api=Vulkan, NVIDIA:DiscreteGPU, #spirv.resource_limits<max_compute_shared_memory_size = 49152, max_compute_workgroup_invocations = 1024, max_compute_workgroup_size = [1024, 1024, 1024], min_subgroup_size = 32, max_subgroup_size = 32, cooperative_matrix_properties_nv = [#spirv.coop_matrix_props<m_size = 8, n_size = 8, k_size = 32, a_type = i8, b_type = i8, c_type = i32, result_type = i32, scope = <Subgroup>>, #spirv.coop_matrix_props<m_size = 16, n_size = 16, k_size = 16, a_type = f16, b_type = f16, c_type = f16, result_type = f16, scope = <Subgroup>>, #spirv.coop_matrix_props<m_size = 16, n_size = 16, k_size = 16, a_type = f16, b_type = f16, c_type = f32, result_type = f32, scope = <Subgroup>>]>>}>
<stdin>:26:3: note: called from
<stdin>:820:10: error: failed to serialize executables
<stdin>:26:3: note: called from
```

### Steps to reproduce your issue

To reproduce the issue, run the following command over the given IR (https://gist.github.com/vivekkhandelwal1/b458ba8e74936b2445693663ff89a967):

```
./build/tools/iree-compile --iree-input-type=tm_tensor --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=vulkan --iree-llvmcpu-embedded-linker-path=/build/compiler/bindings/python/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false --iree-llvmcpu-target-cpu-features=host --iree-stream-resource-index-bits=64 --iree-vm-target-index-bits=64 --iree-vm-bytecode-module-strip-source-map=true -iree-vulkan-target-triple=ampere-unknown-unknown --iree-util-zero-fill-elided-attrs vulkan_reproducible.mlir
```

### What component(s) does this issue relate to?

MLIR, Compiler

**Comment by antiagainst** [comment]
I'm curious what's the Python original input of this dispatch looks like. From what I can see, the following:

```mlir
%23 = linalg.fill ins(%true : i1) outs(%20 : tensor<1x1x1x?xi1>) -> tensor<1x1x1x?xi1>
  %24 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} outs(%22 : tensor<?xi1>) {
  ^bb0(%out: i1):
    %26 = linalg.index 0 : index
    %extracted = tensor.extract %23[%c0, %c0, %c0, %26] : tensor<1x1x1x?xi1>
    linalg.yield %extracted : i1
  } -> tensor<?xi1>
  %25 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} ins(%24 : tensor<?xi1>) outs(%21 : tensor<?xf32>) {
  ^bb0(%in: i1, %out: f32):
    %26 = arith.uitofp %in : i1 to f32
    linalg.yield %26 : f32
  } -> tensor<?xf32>
  flow.dispatch.tensor.store %25, %15, offsets = [0], sizes = [%14], strides = [1] : tensor<?xf32> -> !flow.dispatch.tensor<writeonly:tensor<?xf32>>{%14}
```

`tensor.extract %23` can be folded into `%true`. So `%24` is `%true`, and then `%25` is 1.0. The whole dispatch does not need to exist at all--what's missing is just a pattern to fold `tensor.extract(linalg.fill(%v))` into `%v`.

Is this dispatch a proper translation of the original Python model intent? Wondering whether there is something wrong there.

**Comment by suryajasper** [comment]
@antiagainst I'm still blocked by this issue for the same dispatch, and I've been trying to debug it on my end. Is there anything I can try to resolve or work around it for now while a patch is underway?

**Comment by antiagainst** [comment]
@suryajasper did you read the comment I have in the above?

**Comment by suryajasper** [comment]
Yes, I was able to get the dispatch compiling by folding the tensor.extract op as you mentioned, but I wanted to know the best way to go about implementing this fix for compiling the entire MLIR. Would that involve writing a preprocessing pass to canonicalize every instance of the `tensor.extract(linalg.fill(%x))` pattern, or is there a simpler or more elegant solution I'm overlooking?

**Comment by antiagainst** [comment]
To me what the above pattern is really cleaning up and canonicalize the IR from earlier stages. I think it makes sense to add it as a canonicalization pattern for `linalg.fill` directly, given that this is widely applicable and won't interfere with other functionalities much I believe.

**Comment by suryajasper** [comment]
@antiagainst just opened a PR with the patch I added. I don't have permissions to assign reviewers, so I just wanted to follow up here. Do you think you could take a look and let me know if there's anything I should change or can generalize further?

**Comment by antiagainst** [comment]
Thanks! I guess I wasn't being clear on my suggestions. I commented on the pull request directly. Let me know whether that makes sense! :)

**Comment by suryajasper** [comment]
Wanted to update that I have a PR pending in LLVM with the canonicalization pattern that resolves this issue: https://reviews.llvm.org/D156008

**Comment by antiagainst** [comment]
@suryajasper thanks! commented there. :)

**Comment by suryajasper** [comment]
Thank you! I just revised the PR. I had a test in `canonicalize.mlir` initially, but it didn't show up in the main revision, so I bumped that up as well. It's definitely taking me some time to get used to phabricator 😄

**Comment by suryajasper** [comment]
Hey @antiagainst I was wondering if you had a chance to look over the most recent changes to the LLVM revision. Just following up since there are a few tasks on our side being blocked by the memref issue so it would be great to get the patch in as soon as possible.

**Comment by antiagainst** [comment]
Okay. I saw you still have the patch labelled as "change planned" so it's not in my review queue on https://reviews.llvm.org. Please make sure you request reviews after addressing comments.

**Comment by suryajasper** [comment]
Sorry, didn't realize the label got switched to "changes planned". Just fixed the FileCheck prompts in the test mlir and requested review.

**Comment by suryajasper** [comment]
Squashed the commits on the revision, so now full patch is visible @antiagainst. Let me know if there's anything else that needs changing.

**Comment by antiagainst** [comment]
Landed upstream. I'll close this one given on further work needed other than bring it in with a future LLVM integration.
